server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1020,6 +1021,39 @@ public boolean freeReaderContext(ShardSearchContextId contextId) {
         return false;
     }
 
+    /**
+     * Free reader context if found otherwise return false
+     */
+    public boolean freeReaderContextIfFound(ShardSearchContextId contextId) {
+        try {
+            if (getReaderContext(contextId) != null) {
+                try (ReaderContext context = removeReaderContext(contextId.getId())) {
+                    return context != null;
+                }
+            }
+        } catch (SearchContextMissingException e) {
+            return true;
+        }
+        return true;
+    }
+
+    /**
+     * Free all active pit contexts
+     */
+    public boolean freeAllPitContexts() {
+        final SetOnce<Boolean> isFreed = new SetOnce<>();
```

Comment:
NIT: cant we use simple boolean initialized to true and set to false if free context fails even once? 
This is not a use case for SetOnce.